### PR TITLE
[1] Implement LlamaIndex Tracing

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/12618?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12618/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12618
```

</p>
</details>

### What changes are proposed in this pull request?

Add tracing implementation for LlamaIndex. LLamaIndex recently updated their method of allowing integrators to generate tracing. Previously it was a single callback based, but now it needs two separate **handlers**:

- **SpanHandler** - This handler defines entrypoints for starting and ending spans. Similar to how LangChain tracing works, LlamaIndex already creates their own spans structure and give us those spans and Ids, from which we can build our own MLflow spans.
- **EventHandler** - This handler is primary used for adding supplemental metadata e.g. model name to spans, using `Event` objects that are triggered within the index e.g. LLM start, Agent start, etc.

The above diagram illustrates how these components work together to create a span. Please refer to https://docs.llamaindex.ai/en/stable/module_guides/observability/instrumentation/ for more detailed and accurate explanation🙂

This PR is the first effort to implement the two key **handlers** to generate traces. It actually is 90% of the work to make tracing work, with a few follow-ups to be addressed in the incoming PRs.

**TODOs**
1. Add unit tests for complex models (https://github.com/mlflow/mlflow/pull/12626)
2. (Fully) support streaming and update tests. The handler already mostly works for streaming, but need some minor update like recording chunk events.
3. Add documentation.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: tests/llama_index/test_llama_index_tracer.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+from mlflow.entities.trace_status import TraceStatus
+from mlflow.llama_index.tracer import MlflowEventHandler, MlflowSpanHandler
+from mlflow.tracing.constant import SpanAttributeKey
+
+_TEST_EXPERIMENT = "__test-llama-index-tracing"
+
+
+@pytest.fixture(autouse=True)
+def experiment_id(tmp_path):
+    """Create a new experiment for each test and clean up after."""
```
[the remark this answers is not in the document]
Comment (B-Step62) [reply tests/llama_index/test_llama_index_tracer.py:27]: Two reasons:
1. To get all traces from each test using `search_traces()`, so we can validate there is no unexpected trace. (alternative: use tags to filter per test)
2. Suppress the warning of logging traces to the default experiment.

[the remark this answers is not in the document]
Comment (harupy) [reply tests/llama_index/test_llama_index_tracer.py:27]: Thanks for the explanation. For 1, doesn't `tracking_uri_mock` work?

https://github.com/mlflow/mlflow/blob/859721c63247d21babeff881e35323f38bd713c2/tests/conftest.py#L23-L30

File: mlflow/llama_index/tracer.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        # ignore them for now but they should be accumulated as SpanEvents of the parent span.
+        if isinstance(event, (LLMChatInProgressEvent, LLMCompletionInProgressEvent)):
+            return
+
+        if span := self._span_handler.open_spans.get(event.span_id):
+            self._handle_event(event, span._mlflow_span)
+
+    @singledispatchmethod
+    def _handle_event(self, event: BaseEvent, span: LiveSpan):
+        ...
```
Comment (harupy) [inline mlflow/llama_index/tracer.py:220]:
```suggestion
        raise NotImplementedError
```

should we raise `NotImplementedError` here?

File: tests/llama_index/test_llama_index_tracer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (harupy) [reply tests/llama_index/test_llama_index_tracer.py:27]: maybe I can run this test script without this fixture to see what happens.

File: mlflow/llama_index/tracer.py (hunk 2 above)
Comment (B-Step62) [reply mlflow/llama_index/tracer.py:220]: We don't handle all events, this is no-op impl for events we are not interested in.

File: mlflow/llama_index/tracer.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+            }
+        )
+
+    def _extract_token_counts(self, response) -> Dict[str, int]:
+        if (
+            (raw := getattr(response, "raw", None))
+            and hasattr(raw, "get")
+            and (usage := raw.get("usage"))
+        ):
+            return usage
```
[the remark this answers is not in the document]
Comment (B-Step62) [reply mlflow/llama_index/tracer.py:287]: Either `ChatResponse` or `CompletionResponse`. Let me add type annotations:)

File: tests/llama_index/test_llama_index_tracer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (B-Step62) [reply tests/llama_index/test_llama_index_tracer.py:27]: Oh it actually works (didn't know that fixture auto-applies). Thanks for the catch!!

File: mlflow/llama_index/tracer.py (hunk 2 above)
Comment (harupy) [reply mlflow/llama_index/tracer.py:221]: Can we use `pass`? `...` looks like "forgot to update the code.".

Comment (B-Step62) [reply mlflow/llama_index/tracer.py:221]: Makes sense!

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**